Data structures

lilyminium opened the pull request:
**Universe**
- [x] Universe creation
  - [x] Jupyter notebook for adding topology attributes
- [x] Explaining groups of atoms (residues, segments, fragments, molecules, residuegroups, segment groups)

**AtomGroup**
- [x] AtomGroup creation
- [x] Topology objects
- [ ] Coordinate-related methods

orbeckst commented on doc/source/conf.py:
```diff
@@ -10,9 +10,11 @@
-# sys.path.insert(0, os.path.abspath('.'))
+import os
+import sys
+import MDAnalysis as mda
```
This line will require you to create a full-blown installation (probably via conda) to install MDAnalysis.

Running conda takes about 6 minutes on Travis CI, most of it when it resolves the dependency graph. Perhaps one can speed up by providing a frozen requirements file (or whatever the conda equivalent is).

lilyminium replied:
This is for the API autosummary of code. Possibly that should be done in the MDAnalysis repo itself, like the current docs?
(That's also why there are a million files in this pull request, sorry)

orbeckst replied:
Yes, if it is generated from the docs then it should go into the standard MDA docs. 

Are these API autosummaries crucial for the user guide right now? If yes, then you either need to push hard to get them into MDA develop or make the CI for the UserGuide work.

I think you can use inter-sphinx links to automatically link to the standard docs.

lilyminium replied:
I was arranging the API reference into sections to organise my thoughts (e.g. below) but it makes far more sense to do this in the MDA repo.
<img width="836" alt="Screenshot 2019-10-10 at 9 33 32 AM" src="https://user-images.githubusercontent.com/31115101/66583146-95c4be00-ebce-11e9-81af-6266dcce7cc4.png">

lilyminium commented:
@orbeckst Do you have any thoughts on the current content of the data-structures page: primarily ways to create Universes and AtomGroups, plus links to the Topology section for actions involving:

* adding Residues/Segments
* TopologyObjects
* TopologyAttrs

orbeckst commented on doc/source/data_structures.rst:
```diff
@@ -1,82 +1,516 @@
+Fragments and molecules
+-----------------------
 
+Unlike Residues and Segments, fragments and molecules are not classes in MDAnalysis. A fragment is defined by bond connectivity. A fragment is what is typically considered a molecule: a group of atoms where each atom is bonded to at least one other atom in the fragment, and are not bonded to any atoms outside the fragment. The fragments of a Universe are determined by MDAnalysis as a derived quan ...
```
Not sure what you mean by "are not classes in MDA"? This sounds as if the user needs to understand an implementation detail. Or do you want to make clear that fragments are somewhat different from the standard hierarchy? I think the clearest way to show the difference is by way of examples, as you're doing below.

lilyminium replied:
I took that out. I like the discussion of more detailed developer docs in MDAnalysis/mdanalysis#1175 . Keep the picture Y/N?

![classes](https://user-images.githubusercontent.com/31115101/67347074-c15f8500-f58c-11e9-8f44-e36222976ed6.png)

orbeckst replied:
Keep (with minor modification) – these are basic containers and they have a "is contained in" hierarchical and exclusive relationship, which is really nicely illustrated. But remove "molecule" as a (currently) Gromacs/TPR specific concept. If fragments are not discussed here, say that these are an advanced concept and discussed elsewhere (link).

orbeckst commented on doc/source/data_structures.rst:
```diff
@@ -1,82 +1,516 @@
+    >>> u.atoms.fragments
+    (<AtomGroup with 3341 atoms>,)
 
+In MDAnalysis, a molecule is a GROMACS-only concept. A group of atoms is considered a "molecule" if it is defined by the :code:`[ moleculetype ]` section in a GROMACS topology. Molecules are only defined if a Universe is created from a GROMACS topology. 
```
@jbarnoud @richardjgowers how do we access molecules in the following example?
```python
import MDAnalysis as mda
from MDAnalysisTests import datafiles as data
u = mda.Universe(data.TPR, data.XTC)
```

I see `atoms.molnums` and `atoms.moltypes`. Are the "molecules" the "fragments"? 

Basically, how do use the "molecule" information?

lilyminium replied:
```python
>>> tpr = mda.Universe(TPR)
>>> tpr.atoms.molnums
array([    0,     0,     0, ..., 11086, 11087, 11088])
>>> tpr.atoms.moltypes
array(['AKeco', 'AKeco', 'AKeco', ..., 'NA+', 'NA+', 'NA+'], dtype=object)
```

This is my current understanding: like segments, molecules are read from the TPR file. `moltype` is the name of the molecule given immediately after the `[ moleculetype ]` header in the [human-readable topology files](http://manual.gromacs.org/archive/5.0/online/top.html). `molnum` is the order of the molecules in the `[ molecules ]` section. They're only available for universes created from tpr files; gro files don't include this information. Some methods can be applied to molecules. There's relevant discussion in #2 .

Fragments are MDAnalysis derived groups of atoms that are connected by bonds. They're available for all formats. Unsure what happens when you load in a file without connectivity information and then `guess_bonds` later -- do Fragments get re-computed?

A better name for `moltypes` might be `molnames` because GROMACS already has a molecule type concept with `residuetypes.dat`, which is a mapping of molecule names to Protein/DNA/RNA/water/ion categories.

orbeckst replied:
So the `molnum`/`moltypes` are just attributes and do not create any objects. Then it should definitely not show up in the figure.

> A better name for moltypes might be molnames because GROMACS already has a molecule type concept with residuetypes.dat, which is a mapping of molecule names to Protein/DNA/RNA/water/ion categories.

Yes, I agree, especially as "moleculetype" is the category. Maybe "moleculetype.name" ;-). (No)

orbeckst commented on doc/source/index.rst:
```diff
@@ -24,5 +24,8 @@ Link to gallery.
 
    data_structures
+   topology_system
+   coordinate_transformation
```
I wouldn't call it "coordinate_transformation" because it gets really confusing with the "trajectory transformations" and the "mathematical transformations". 

I understand that we are transforming coordinates into something else but perhaps adapt the title "working with coordinates" for the filename, too?

orbeckst commented on doc/source/selections.rst:
```diff
@@ -0,0 +1,440 @@
+    ``prop z >= 5.0`` selects all atoms with z coordinate greater than 5.0;
+    ``prop abs z <= 5.0`` selects all atoms within -5.0 <= z <= 5.0.
+
+From version 0.6 onwards, some geometric selections (around, sphlayer,
```
This section is outdated. Need to be clearer about periodicity handling – probably needs tobe linked to a whole page on this topic.

lilyminium replied:
I skimmed a few issues and some of the code. It looks like:

* a box is passed into every geometric keyword
* selection only uses the bruteforce or nsgrid methods for distances
* periodic=True by default if there's a box (and nsgrid creates a pseudo box if not, but brute force doesn't seem to?)
* because of the automatic method choosing, it's all reasonably fast
* and works for any triclinic box?

orbeckst commented on doc/source/selections.rst:
```diff
@@ -0,0 +1,440 @@
+periodic boundary conditions should be taken into account then change the
+default behaviour of MDAnalysis by setting these two flags::
+
+  MDAnalysis.core.flags['use_periodic_selections'] = True
```
remove – is about to be killed https://github.com/MDAnalysis/mdanalysis/issues/782

orbeckst commented on doc/source/selections.rst:
```diff
@@ -0,0 +1,440 @@
+Index
+-----
+
+bynum *index-range*
```
We need to rewrite this and not call it "index" but "atom number" – anything different from "index". In reality it just adds 1 to the 0-based index. It would be nice if we were to use it for "atom numbers" obtained from the topology file, such as PDB AtomID or TPR index https://github.com/MDAnalysis/mdanalysis/issues/2364

orbeckst commented on doc/source/selections.rst:
```diff
@@ -0,0 +1,440 @@
+    :class:`MDAnalysis.Universe` are consecutively numbered, and the index
+    runs from 1 up to the total number of atoms.
+
+index *index-range*
```
This directly uses the MDAnalysis unique atom index `u.atoms.ix`

orbeckst commented on doc/source/selections.rst:
```diff
@@ -0,0 +1,440 @@
+ >>> static_ag
+ <UpdatingAtomGroup with 3 atoms>
+
+.. _instance-selectors:
```
Remove the whole section on instant selectors. https://github.com/MDAnalysis/mdanalysis/issues/1377

orbeckst commented on doc/source/topology_system.rst:
```diff
@@ -0,0 +1,89 @@
+
+To add a :code:`Residue` or :code:`Segment` to a topology, use the :code:`Universe.add_Residue` or :code:`Universe.add_Segment` methods.
+
+.. code-block::
```
The example is ok but better would be one of the FAQs: "How do I create different "chains" for protein consisting of multiple polypeptides?" (e.g., when loaded from a GRO file) or how do I create custom segments for "protein", "lipids", "solvent"?

orbeckst replied:
https://www.mdanalysis.org/docs/documentation_pages/topology/init.html#supported-topology-formats

lilyminium replied:
There's now a section in `constructing_universe.ipynb` that creates custom segments for protein domains. I could also make a mini-notebook of this, if that doesn't seem redundant.

lilyminium replied:
Also, moved it to the Groups of atoms page.

orbeckst commented on doc/source/topology_system.rst:
```diff
@@ -0,0 +1,89 @@
+
+Several topology attributes are central to MDAnalysis, and are derived for each Universe regardless of format:
+
+
```
@richardjgowers @jbarnoud – what is the minimum list of topology attributes?

On that note, is there any chance that we get the docs back for transplanted methods that are NOT generated by default? https://github.com/MDAnalysis/mdanalysis/pull/2176

lilyminium replied:
I put my notes on [the UserGuide wiki here.](https://github.com/MDAnalysis/UserGuide/wiki/Topology-attributes) It looks like every format will give the following topology attributes:

* atom id
* atom index
* res id / res num
* res index
* seg id
* seg index
* types (alternately atom types or elements)
* masses (guessed)

orbeckst replied:
Thanks. I added a link in https://github.com/MDAnalysis/mdanalysis/issues/2364#issuecomment-546455039

orbeckst commented on doc/source/topology_system.rst:
```diff
@@ -0,0 +1,89 @@
+Adding or modifying TopologyAttrs
+---------------------------------
+
+See this notebook on colouring a protein by RMSF with the tempfactor attribute.
```
link?

orbeckst commented on doc/source/topology_system.rst:
```diff
@@ -0,0 +1,89 @@
+
+TopologyObjects can only be read from these file formats:
+
+    * PSF
```
- TPR
- PDB with CONECT (I think, at least bonds)
- AMBER prmtop
- DMS
- MOL2
- LAMMPS DATA
- Tinker TXYZ
- HoomdXML
- MMTF

All these formats generate at least bonds.

lilyminium commented on doc/source/data_structures.rst:
```diff
@@ -1,82 +1,516 @@
 -------------------------
 
-Universe.empty()
+A Universe can be constructed from scratch with :func:`Universe.empty`. There are three stages to this process:
```
This did the trick:

`:meth:Universe.empty <MDAnalysis.core.universe.Universe.empty>`

lilyminium commented on doc/source/data_structures.rst:
```diff
@@ -1,82 +1,516 @@
+    >>> u.atoms.groupby(['masses', 'resnames'])
+    {(32.06, 'MET'): <AtomGroup with 6 atoms>, (32.06, 'CYS'): <AtomGroup with 1 atom>, (1.008, 'LEU'): <AtomGroup with 176 atoms>, (1.008, 'SER'): <AtomGroup with 25 atoms>, (1.008, 'TYR'): <AtomGroup with 63 atoms>, (1.008, 'ARG'): <AtomGroup with 169 atoms>, (1.008, 'GLU'): <AtomGroup with 108 atoms>, (1.008, 'GLY'): <AtomGroup with 60 atoms>, (1.008, 'ASN'): <AtomGroup with 24 atoms>, ..., }
+
+Constructing from Atoms
```
I added a section under Order and Uniqueness.

lilyminium commented:
@orbeckst Unless you have any concerns I might merge this sometime tomorrow (Australian).

richardjgowers commented on .travis.yml:
```diff
@@ -17,9 +17,9 @@ before_install:
+- jupyter-nbextension enable nglview --py --sys-prefix
 script:
 - make -C ${SPHINX_DIR} html && touch ${HTML_DIR}/.nojekyll
 deploy:
```
Kill the DS Store file below

orbeckst replied:
Add a top-level `.gitignore` ! You can copy the MDA [.gitignore](https://github.com/MDAnalysis/mdanalysis/blob/develop/.gitignore)

richardjgowers commented on doc/source/atomgroup.rst:
```diff
@@ -0,0 +1,331 @@
+    >>> atom1 = u.atoms[4]
+    >>> atom2 = u.atoms[6]
+    >>> atom3 = u.atoms[2]
+    >>> ag = mda.AtomGroup([atom1, atom2, atom3])
```
Maybe a cute hack to mention is you can use the built in sum function

lilyminium replied:
I didn't know you could do that -- that is cute!

richardjgowers commented on doc/source/groups_of_atoms.rst:
```diff
@@ -0,0 +1,117 @@
+
+.. code-block::
+
+    >>> u.atoms.fragments
```
Can we use a better U here? This doesn’t actually show the thing in question

orbeckst replied:
The TPR/XTC has protein, water, ions. We also have a small peptide in a membrane (PEPT?)

orbeckst replied:
Or KALP – I am not 100% sure that we have that as a data file.

lilyminium replied:
I couldn't find the small peptide in a membrane, so I used the lysozyme in water (which has easily recognisable fragments)

orbeckst commented on doc/source/groups_of_atoms.rst:
```diff
@@ -0,0 +1,117 @@
+
+.. image:: images/classes.png
+
+First and foremost is the :class:`~MDAnalysis.core.groups.AtomGroup`. An :class:`~MDAnalysis.core.groups.AtomGroup` is the primary :class:`~MDAnalysis.core.groups.Atom` container; virtually everything can be accessed through it, as detailed `above <https://www.mdanalysis.org/UserGuide/data_structures.html#atomgroup>`_. This includes chemically meaningful groups of :class:`~MDAnalysis.core.groups. ...
```
Don't use the inline link for "above", use a sphinx ref, which will never break.

orbeckst commented on doc/source/groups_of_atoms.rst:
```diff
@@ -0,0 +1,117 @@
+Residues and Segments
+---------------------
+
+A :class:`~MDAnalysis.core.groups.Residue` is composed of :class:`~MDAnalysis.core.groups.Atom`\ s, and a :class:`~MDAnalysis.core.groups.Segment` is composed of :class:`~MDAnalysis.core.groups.Residue`\ s.
```
Do the plurals render nicely? (I have to remember the backslash trick, I used to reformulate as "Atom instances"...)

lilyminium replied:
As well as it can.

<img width="674" alt="Screenshot 2019-10-26 at 2 16 09 PM" src="https://user-images.githubusercontent.com/31115101/67613818-929c1580-f7fd-11e9-9bf1-405fc1a38f72.png">

You can also change the text with `Atoms <MDAnalysis.core.groups.Atom>`. Is this too misleading?

<img width="668" alt="Screenshot 2019-10-26 at 2 17 37 PM" src="https://user-images.githubusercontent.com/31115101/67613825-b5c6c500-f7fd-11e9-9780-c02f730cec40.png">

Or probably not every plural has to be a class link. (I will change those lingering :code:Residues etc. to classes!)

orbeckst replied:
I wouldn't change the text. We also singular/plural attributes. Better to use the formatting to make clear what the actual name is.

orbeckst commented on doc/source/groups_of_atoms.rst:
```diff
@@ -0,0 +1,117 @@
+
+A :class:`~MDAnalysis.core.groups.Residue` is composed of :class:`~MDAnalysis.core.groups.Atom`\ s, and a :class:`~MDAnalysis.core.groups.Segment` is composed of :class:`~MDAnalysis.core.groups.Residue`\ s.
+
+These are considered atom :ref:`topology attributes <topology-attributes>`.
```
This sentence might be true but is confusing unless you are someone who has had nightmares over topology attributes and what they are... I would move it further down, here it seems to distract from the "container" aspect.

orbeckst replied:
Or use a footnote.

orbeckst commented on doc/source/groups_of_atoms.rst:
```diff
@@ -0,0 +1,117 @@
+    >>> a.residue.segment.residues
+    <ResidueGroup with 214 residues>
+
+Adding to a topology
```
Although this fits here, it's less important for someone starting out. I know it's hard to group things. Perhaps leverage the fact that you're writing hyperlinked docs and put it in a different section and link to it?

orbeckst replied:
The thrust of the previous section seemed to be on accessing and "reading" not on creating and "writing" topologies.

lilyminium replied:
Good idea, there is already a link to a relevant section in a notebook below as well.

lilyminium replied:
Moved this and Molecules to advanced_topology.rst for now

orbeckst commented on doc/source/groups_of_atoms.rst:
```diff
@@ -0,0 +1,117 @@
+    >>> u.atoms.fragments
+    (<AtomGroup with 3341 atoms>,)
+
+Molecules
```
"Molecules" is another "advanced" section that distracts from the important things. I would put it in an advanced topolgy chapter and footnote/link.

lilyminium replied:
Fair enough. "What is a molecule?" was one of the immediate questions I had when I was looking at AtomGroup's split, accumulate, center/centroid methods (and "why isn't it in the other methods?") but hyperlinks and the search bar hopefully make this a non-issue.

orbeckst commented on doc/source/selections.rst:
```diff
@@ -0,0 +1,344 @@
+
+AtomGroups can be created by selecting atoms using the MDAnalysis atom selection language:
+
+  >>> kalp = u.select_atoms("segid KALP")
```
I am not sure that we have KALP as a data file. Let's use something that we have and can immediately use.

I imagine keen readers to copying and pasting in their own ipython/notebooks to try out what you have.

orbeckst replied:
Btw, KALP is a relic of @nmichaud 's original project for which he wrote MDAnalysis.

orbeckst commented on doc/source/selections.rst:
```diff
@@ -0,0 +1,344 @@
+:class:`~MDAnalysis.core.groups.AtomGroup`. Selections always return an
+:class:`~MDAnalysis.core.groups.AtomGroup` with atoms sorted according to their
+index in the topology. This is to ensure that there are not any duplicates,
+which can happen with complicated selections.
```
Say here that you start with a number of examples and then explain more later. Otherwise this is a bit much for a newbie.

orbeckst replied:
I do like the use of links for anyone wanting to learn more, I think that's really helpful.

orbeckst commented on doc/source/selections.rst:
```diff
@@ -0,0 +1,344 @@
+Simple selections
+-----------------
+
+protein
```
Perhaps format the selection keyword in a particular way? Bold?

lilyminium replied:
This is how it renders in ReadTheDocs for me, is that what you had in mind? 

<img width="677" alt="Screenshot 2019-10-26 at 2 44 58 PM" src="https://user-images.githubusercontent.com/31115101/67613918-3a661300-f7ff-11e9-9df5-289add62d5a4.png">

orbeckst replied:
That's fine, I didn't realize that the list would render it differently.

orbeckst commented on doc/source/standard_selections.rst:
```diff
@@ -0,0 +1,95 @@
+
+The below names are drawn from the CHARMM 27, OPLS-AA, GROMOS 53A6, AMBER 03, and AMBER 99sb*-ILDN force fields.
+
++------+------+------+------+------+------+------+------+
```
Is there a way to generate this table programmatically from the MDAnalysis version that the UserGuide pertains to?

Although this is not likely to change soon, when it does I guarantee that someone will forget to update the UserGuide.

I think @jbarnoud has written code to, for instance, dynamically create the AUTHORS list in the docs.

lilyminium replied:
I think that sticking it in `conf.py` to run automatically would require installing MDAnalysis. If that isn't too heavy, this would be quite easy. Otherwise a script could be run manually from time-to-time to just regenerate `standard_selections.rst`?

jbarnoud replied:
At least, you can have the table in a separate file that gets included here. This way, you can generate the table with a script without having to deal with the rest of the page in the script. I have an example in <https://github.com/MDAnalysis/mdanalysis/pull/2176>; it is a bit of a monster, but mostly because of what gets written in the external file.

orbeckst commented on doc/source/standard_selections.rst:
```diff
@@ -0,0 +1,95 @@
+The residue names listed here are accessible via the "nucleic" keyword in the :ref:`selections`. 
+
+The below names are drawn from largely from the CHARMM force field.
+
```
Is there a way to generate this table programmatically from the MDAnalysis version that the UserGuide pertains to?

Although this is not likely to change soon, when it does I guarantee that someone will forget to update the UserGuide.

I think @jbarnoud has written code to, for instance, dynamically create the AUTHORS list in the docs.

orbeckst commented on doc/source/standard_selections.rst:
```diff
@@ -0,0 +1,95 @@
+Nucleobases
+-----------
+
+Nucleobase atoms from nucleic acid residues are recognised based on their names in CHARMM.
```
Is there a way to generate this table programmatically from the MDAnalysis version that the UserGuide pertains to?

Although this is not likely to change soon, when it does I guarantee that someone will forget to update the UserGuide.

I think @jbarnoud has written code to, for instance, dynamically create the AUTHORS list in the docs.

orbeckst commented on doc/source/selections.rst:
```diff
@@ -0,0 +1,344 @@
+
+    MDAnalysis will ignore everything after the ``*``. ``u.select_atoms("resname *E")`` will not select atoms whose residue name ends in E, but instead select every atom.
+
+
```
Is there a way that we only maintain this nice documentation of selections in a single place? If we update it in the code then we mustn't forget to update it here. (Not something to be solved in this PR but generally something to think about: How do we avoid duplicating docs that need to be updated.)

orbeckst commented on doc/source/trajectories.rst:
```diff
@@ -0,0 +1,7 @@
+Trajectories
+===============
+
+MDAnalysis groups dynamic data about a :code:`Universe` into its trajectory. This is typically loaded from a trajectory file.
```
TBD ;-)

lilyminium commented:
Thanks for the reviews @richardjgowers and @orbeckst, and for the pointer to the script @jbarnoud. I have now:

* switched the KALP fragment example to lysozyme in water
* updated the constructing universes notebook to have a correct water geometry and added a section on tiling
*  I hope removed .DS_Store?
* added a script to generate files for the tables in `standard_selections.rst`

@orbeckst NGLView is cool! You can save the notebook widget state, and I think I had to include the javascript for widgets with this for it to render:

```
html_js_files = [
    'https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js',
    DEFAULT_EMBED_REQUIREJS_URL,
]
```

([from this issue](https://github.com/spatialaudio/nbsphinx/issues/84))

re: the MDAnalysis version -- I think ideally the UserGuide should apply to the most up-to-date one, with notes on when major changes happened? The tutorials should definitely have minimum versions on them. I haven't added them yet, but will once I work out which versions they work with.

orbeckst commented:
Sorry, my comments are too late to the party – but they were all basically 👍 .
